orchagent/policerorch.cpp

```diff
@@ -123,119 +142,395 @@ void PolicerOrch::doTask(Consumer &consumer)
+            }
+            /*CIR is mandatory parameter*/
+            if (!cir)
+            {
+                SWSS_LOG_ERROR("Failed to create storm control policer %s,\
+                        missing madatory fields", storm_policer_name.c_str());
+                /*
+                 * return here as the same error 
+                 * would be seen for the complete iteration
+                 */
+                return task_process_status::task_invalid_entry;
+            }
 
-                attrs.push_back(attr);
+            /*Enabling storm-control on port*/
+            sai_attribute_t port_attr;
+            if (storm_type == storm_broadcast)
+            {
+                port_attr.id = SAI_PORT_ATTR_BROADCAST_STORM_CONTROL_POLICER_ID;
+            }
+            else if (storm_type == storm_unknown_unicast)
+            {
+                port_attr.id = SAI_PORT_ATTR_FLOOD_STORM_CONTROL_POLICER_ID;
+            }
+            else if (storm_type == storm_unknown_mcast)
+            {
+                port_attr.id = SAI_PORT_ATTR_MULTICAST_STORM_CONTROL_POLICER_ID;
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Unknown storm_type %s", storm_type.c_str());
+                /*continue as there can be more interfaces*/
+                continue;
             }
 
+			sai_object_id_t policer_id;
             // Create a new policer
             if (!update)
-            {
-                if (!meter_type || !mode)
-                {
-                    SWSS_LOG_ERROR("Failed to create policer %s,\
-                            missing mandatory fields", key.c_str());
-                }
-
-                sai_object_id_t policer_id;
-                sai_status_t status = sai_policer_api->create_policer(
-                    &policer_id, gSwitchId, (uint32_t)attrs.size(), attrs.data());
-                if (status != SAI_STATUS_SUCCESS)
-                {
-                    SWSS_LOG_ERROR("Failed to create policer %s, rv:%d",
-                            key.c_str(), status);
-                    it++;
-                    continue;
-                }
-
-                SWSS_LOG_NOTICE("Created policer %s", key.c_str());
-                m_syncdPolicers[key] = policer_id;
-                m_policerRefCounts[key] = 0;
-            }
+			{
+				sai_status_t status = sai_policer_api->create_policer(
+						&policer_id, gSwitchId, (uint32_t)attrs.size(), attrs.data());
+				if (status != SAI_STATUS_SUCCESS)
+				{
+					SWSS_LOG_ERROR("Failed to create policer %s, rv:%d",
+							storm_policer_name.c_str(), status);
+					/* 
+					 * Returning here since the same failure 
+					 * would be seen for the full iteration
+					 */
+					return task_process_status::task_need_retry;
+				}
+
+				SWSS_LOG_NOTICE("Created storm-control policer %s", storm_policer_name.c_str());
+                m_syncdPolicers[storm_policer_name].policerOid = policer_id;
+				m_policerRefCounts[storm_policer_name] = 0;
```

Comment:
This design is based on the configuration syntax used. Industry standard configuration doesn't use a policer template and instead the policer parameters (cir/cbs,...) are configured directly. There is no named policer in case of BUM storm-control configuration. Hence one policer is created per configuration on interface. 
Policer templates can be added as an enhancement where the configuration syntax can look as - "config interface storm-control broadcast add <POLICER_TEMPLATE_NAME>. 
Please suggest.

Reply:
This question is more towards the usage of hardware resources if we want to use this feature to drop all unknown unicast, unknown-mcast, broadcast storm on all ports. What are your thoughts around that? What is the limit on the number of internal policers that can be created?

Reply:
In worst case scenario, num_policers_required = num_phy_interfaces * 3.
One point to support creation of separate policers for every storm-control configuration - When user wants to rate-limit to 1000 kbps on a set of interfaces (eg. Ethernet1 - Ethernet10). In case only one policer is created in hardware, the ingress traffic on all the 10 interfaces would be rate-limited by the same policer in hardware. In this case, Orch should create a policer with 1000 kbps *10 as the actual CIR value which might result in incorrect rate-limiting in case there is excessive traffic on few ports and less traffic on other ports. 
When statistics support is required, it would be feasible if there are separate policers per configuration. 
Please provide your inputs on possible alternatives.

Reply:
Can you please comment on the hw resource usage for the worst case (num_phy_interfaces * 3)? Also what is the max limit on the number of internal policers that can be created?

Reply:
Depending on hardware capability, following are the possibilities of support. 
- Hardware has dedicated storm policer resource per interface. 
- Hardware doesn't have dedicated policers. Hence policer resource is allotted from a shared pool. 
In scenario1, there is no constraint w.r.t hardware resource usage as there are dedicated policers per port. 
In scenario2, scaled configuration, 3 policers are required per physical interface multiplied by maximum physical ports available in the system.